### PR TITLE
Adding AuthorizedScopesHandler to Publish Authorized Identities From EdgeHub to Broker

### PR DESCRIPTION
**Pull request by dylanbronson**
Adding AuthorizedScopesHandler, which implements IMessageProducer.
AuthorizedScopesHandler will send all authorized identities when the OnConnect event is triggered by IMqttBrokerConnector (on first connect).

It will also send all authorized identities every time the ServiceIdentitiesUpdated event is triggered by IDeviceScopeIdentitiesCache.

We are sending the serialized list of authorized identities on the topic "$edgehub/authorization," which will be received by the MqttBroker

### CODE REVIEW

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            }
+            else
+            {
+                this.brokerServiceIdentities.Add(brokerServiceIdentity.Identity, brokerServiceIdentity);
+            }
+        }
+
+        async Task PublishAddOrUpdateBrokerServiceIdentities(Dictionary<string, BrokerServiceIdentity> brokerServiceIdentities)
+        {
+            foreach (BrokerServiceIdentity brokerServiceIdentity in brokerServiceIdentities.Values)
```
Comment (vadim-kovalyov) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:71]: can we batch send the identities instead? For individual Add/Update we can send an array of 1 element.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity.Service;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    public class ScopeIdentitiesHandler : IMessageProducer
+    {
+        const string AddOrUpdateTopic = "$internal/identities/addOrUpdate";
+        const string RemoveTopic = "$internal/identities/removeTopic";
```
Comment (vadim-kovalyov) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:19]: nit:
```suggestion
        const string RemoveTopic = "$internal/identities/remove";
```

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Core.Test/DeviceScopeIdentitiesCacheTest.cs [hunk 3]
```diff
@@ -671,6 +641,27 @@ public async Task GetAuthChainTest()
...
+            // Arrange
+            var store = GetEntityStore("cache");
+            var serviceIdentityHierarchy = new Mock<IServiceIdentityHierarchy>();
+            serviceIdentityHierarchy.Setup(s => s.GetAllIds()).ReturnsAsync(new List<string> { "d1", "d2", "d3" });
+            var serviceProxy = new Mock<IServiceProxy>();
+
+            var deviceScopeIdentitiesCache = await DeviceScopeIdentitiesCache.Create(serviceIdentityHierarchy.Object, serviceProxy.Object, store, TimeSpan.FromHours(1));
+
+            // Act
+            var ids = await deviceScopeIdentitiesCache.GetAllIds();
```
Comment (richma-ms) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Core.Test/DeviceScopeIdentitiesCacheTest.cs:656]: >GetAllIds(); [](start = 55, length = 12)

I'm not seeing this on the interface definition?

**Comment by richma-ms** [comment]
>             bool hasUpdated = existing.HasValue && !existing.Contains(serviceIdentity);

upon closer inspection of this logic, I don't think the event fires for newly added identities, you might need to update this (depending on expectation of existing callers)

---
Refers to: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs:244 in 55e7b45. [](commit_id = 55e7b455e22e58f010eeb957cad8e0d9bbc2a49f, deletion_comment = False)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        readonly IDeviceScopeIdentitiesCache deviceScopeIdentitiesCache;
+        readonly AtomicBoolean connected = new AtomicBoolean(false);
+        IMqttBrokerConnector connector;
+        Dictionary<string, BrokerServiceIdentity> brokerServiceIdentities = new Dictionary<string, BrokerServiceIdentity>();
+
+        public ScopeIdentitiesHandler(IDeviceScopeIdentitiesCache deviceScopeIdentitiesCache)
+        {
+            this.deviceScopeIdentitiesCache = deviceScopeIdentitiesCache;
+            this.deviceScopeIdentitiesCache.ServiceIdentityUpdated += (sender, serviceIdentity) => Task.WhenAll(this.ServiceIdentityUpdated(serviceIdentity));
+            this.deviceScopeIdentitiesCache.ServiceIdentityRemoved += (sender, identity) => Task.WhenAll(this.ServiceIdentityRemoved(identity));
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:30]: this is the race concern that I brought up this morning.  The refresh starts when the cache object is instantiated, and an indeterminable amount of time has passed between and you hooking up the delegates here.  There's no guarantee that you won't be missing callbacks.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs [hunk 5]
```diff
@@ -107,10 +89,14 @@ await serviceIdentity
             {
                 Events.ErrorRefreshingCache(e, id);
             }
+        }
 
-            if (invokeServiceIdentitiesUpdated)
+        public async Task RefreshServiceIdentities(IEnumerable<string> ids)
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs:94]: >RefreshServiceIdentities [](start = 26, length = 24)

my neg-caching PR went in, I think you're gonna get merge conflicts here.  Lemme know if you need help resolving.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs [hunk 6]
```diff
@@ -125,10 +125,10 @@ public async Task RefreshServiceIdentities(IEnumerable<string> ids)
                 await this.RefreshServiceIdentityInternal(id, false);
             }
 
-            this.ServiceIdentitiesUpdated?.Invoke(this, await this.serviceIdentityHierarchy.GetAllIds());
+            this.ServiceIdentitiesUpdated?.Invoke(this, await this.serviceIdentityHierarchy.GetAllServiceIdentities());
         }
 
-        async Task RefreshServiceIdentityInternal(string id, bool invokeServiceIdentitiesUpdated)
+        public async Task RefreshServiceIdentityInternal(string id, bool invokeServiceIdentitiesUpdated)
```
Comment (vadim-kovalyov) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs:131]: q: do we need this to be public?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/BrokerServiceIdentity.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Json;
+    using Newtonsoft.Json;
+
+    public class BrokerServiceIdentity : IComparable
```
Comment (vadim-kovalyov) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/BrokerServiceIdentity.cs:9]: nit: I would add some code comments where this DTO is used.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    using System.Text;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity.Service;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    public class ScopeIdentitiesHandler : IMessageProducer
```
Comment (vadim-kovalyov) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:17]: nit: class comments.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        {
+            this.deviceScopeIdentitiesCache = deviceScopeIdentitiesCache;
+            this.deviceScopeIdentitiesCache.ServiceIdentitiesUpdated +=
+                async (sender, serviceIdentities) => await this.ServiceIdentitiesUpdated(serviceIdentities);
+        }
+
+        public void SetConnector(IMqttBrokerConnector connector)
+        {
+            this.connector = connector;
+            this.connector.OnConnected += async (sender, args) => await this.OnConnect();
```
Comment (vadim-kovalyov) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:35]: :+1:

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        async Task OnConnect()
+        {
+            this.connected.Set(true);
+            await this.PublishBrokerServiceIdentities(this.lastBrokerServiceIdentityUpdate);
+        }
+
+        async Task ServiceIdentitiesUpdated(IList<ServiceIdentity> serviceIdentities)
+        {
+            IList<BrokerServiceIdentity> brokerServiceIdentities =
+                    await Task.WhenAll(
```
Comment (vadim-kovalyov) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:47]: I'm not sure if running `GetAuthChain` in parallel will provide a benefit. AFAIK, `ServiceIdentityTree` implementation uses nodeLock inside, so each call will be exclusive anyways. 

If you want to make it fast, maybe we should add a overloaded method to `IServiceIdentityHierarchy` that accepts list of ids and returns the auth chains.

@richma-ms do you have any thoughts on that?

Comment (dylanbronson) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:47]: I changed it to just be a loop that calls GetAuthChain on every Id. We can make it marginally faster if we want to add that extra method to IServiceIdentityHierarchy

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs [hunk 11]
```diff
@@ -144,6 +144,10 @@ await serviceIdentity
                     // don't repeatedly refresh the same identity
                     // in rapid succession.
                     this.identitiesLastRefreshTime[id] = DateTime.UtcNow;
+                    if (invokeServiceIdentitiesUpdated)
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs:147]: >if (invokeServiceIdentitiesUpdated) [](start = 20, length = 35)

nit: add blank line before the if-statement here as the comment doesn't apply

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs (hunk 10 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs]: I don't think it'll make a noticeable difference, authchain is calculcated and cached on identity insertion/updation, so GetAuthChain is just a constant-time look up.

---
In reply to: [477477814](https://github.com/Azure/iotedge/pull/3439#discussion_r477477814) [](ancestors = 477477814)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+            await this.PublishBrokerServiceIdentities(this.lastBrokerServiceIdentityUpdate);
+        }
+
+        async Task ServiceIdentitiesUpdated(IList<ServiceIdentity> serviceIdentities)
+        {
+            IList<BrokerServiceIdentity> brokerServiceIdentities = new List<BrokerServiceIdentity>();
+            foreach (ServiceIdentity serviceIdentity in serviceIdentities)
+            {
+                brokerServiceIdentities.Add(
+                    new BrokerServiceIdentity(serviceIdentity.Id, await this.deviceScopeIdentitiesCache.GetAuthChain(serviceIdentity.Id)));
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:53]: >serviceIdentity.Id [](start = 46, length = 18)

why get the whole ServiceIdentity from the cache if you're just using the id?  Isn't that already exposed by GetAllIds?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+        public void SetConnector(IMqttBrokerConnector connector)
+        {
+            this.connector = connector;
+            this.connector.OnConnected += async (sender, args) => await this.OnConnect();
+        }
+
+        async Task OnConnect()
+        {
+            this.connected.Set(true);
+            await this.PublishBrokerServiceIdentities(this.lastBrokerServiceIdentityUpdate);
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:44]: >this.lastBrokerServiceIdentityUpdate [](start = 54, length = 36)

I don't see anything guaranteeing this is populated at OnConnect time.  Is it going for the list to be empty if ServiceIdentitiesUpdated fires later than OnConnect?

Comment (dylanbronson) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:44]: There's a check at the beginning of PublishBrokerServiceIdentities that will do a no-op if the list is empty

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            this.connected.Set(true);
+            if (this.lastBrokerServiceIdentityUpdate.Count == 0)
+            {
+                this.lastBrokerServiceIdentityUpdate =
+                    await this.ConvertIdsToBrokerServiceIdentities(await this.deviceScopeIdentitiesCache.GetAllIds());
+            }
+
+            await this.PublishBrokerServiceIdentities(this.lastBrokerServiceIdentityUpdate);
+            this.lastBrokerServiceIdentityUpdate.Clear();
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs:50]: >this.lastBrokerServiceIdentityUpdate.Clear(); [](start = 12, length = 45)

seems unnecessary, right?  You don't clear these in ServiceIdentitiesUpdated.  Can prob just let it be overwritten.

**Pull request merged**